### PR TITLE
Enabled multiple join implementations in CBO

### PR DESCRIPTION
**Pull request by pavelvelikhov**
- Enabled multiple join implementations in CBO

### Changelog entry 

...

### Changelog category 

* New feature
* Experimental feature

### Additional information

...

### CODE REVIEW

File: ydb/library/yql/core/cbo/cbo_optimizer_new.h [hunk 1]
```diff
@@ -74,6 +79,54 @@ enum EJoinKind: ui32
...
 
+/**
+ * This is a temporary structure for KQP provider
+ * We will soon be supporting multiple providers and we will need to design
+ * some interfaces to pass provider-specific context to the optimizer
+*/
+struct IProviderContext {
+    virtual ~IProviderContext() = default;
+
+    virtual double ComputeJoinCost(const TOptimizerStatistics& leftStats, const TOptimizerStatistics& rightStats, EJoinImplType joinImpl) const = 0;
```
Comment (resetius) [inline ydb/library/yql/core/cbo/cbo_optimizer_new.h:90]: I would rename it to smth like EJoinAlgoType. Don't like this `Impl` in naming.

File: ydb/library/yql/core/cbo/cbo_optimizer_new.h [hunk 2]
```diff
@@ -74,6 +79,54 @@ enum EJoinKind: ui32
...
+};
+
+/**
+ * Temporary solution for default provider context
+*/
+
+struct TDummyProviderContext : IProviderContext {
+    TDummyProviderContext() {}
+
+    double ComputeJoinCost(const TOptimizerStatistics& leftStats, const TOptimizerStatistics& rightStats, EJoinImplType joinImpl) const {
```
Comment (resetius) [inline ydb/library/yql/core/cbo/cbo_optimizer_new.h:106]: Pls add override

File: ydb/library/yql/core/cbo/cbo_optimizer_new.h [hunk 3]
```diff
@@ -74,6 +79,54 @@ enum EJoinKind: ui32
...
+        Y_UNUSED(joinImpl);
+        return leftStats.Nrows + 2.0 * rightStats.Nrows;
+    }
+
+    bool IsJoinApplicable(const std::shared_ptr<IBaseOptimizerNode>& left,
```
Comment (resetius) [inline ydb/library/yql/core/cbo/cbo_optimizer_new.h:111]: Pls add override

File: ydb/library/yql/core/cbo/cbo_optimizer_new.h [hunk 4]
```diff
@@ -74,6 +79,54 @@ enum EJoinKind: ui32
...
+        const std::set<std::pair<NDq::TJoinColumn, NDq::TJoinColumn>>& joinConditions,
+        EJoinImplType joinImpl) = 0;
+
+};
+
+/**
+ * Temporary solution for default provider context
+*/
+
+struct TDummyProviderContext : IProviderContext {
```
Comment (resetius) [inline ydb/library/yql/core/cbo/cbo_optimizer_new.h:103]: public IProviderContext

File: ydb/library/yql/core/cbo/cbo_optimizer_new.h [hunk 5]
```diff
@@ -74,6 +79,54 @@ enum EJoinKind: ui32
...
+
+        Y_UNUSED(left);
+        Y_UNUSED(right);
+        Y_UNUSED(joinConditions);
+        Y_UNUSED(joinImpl);
+
+        return true;
+    }
+
+    static const TDummyProviderContext& instance() {
```
Comment (resetius) [inline ydb/library/yql/core/cbo/cbo_optimizer_new.h:124]: Why do we need singleton here?

**Review by resetius — DISMISSED** [review]
Pls fix my comments in next PR

File: ydb/library/yql/core/cbo/cbo_optimizer_new.h (hunk 1 above)
Comment (pavelvelikhov) [reply ydb/library/yql/core/cbo/cbo_optimizer_new.h:90]: done

File: ydb/library/yql/core/cbo/cbo_optimizer_new.h (hunk 2 above)
Comment (pavelvelikhov) [reply ydb/library/yql/core/cbo/cbo_optimizer_new.h:106]: done

File: ydb/library/yql/core/cbo/cbo_optimizer_new.h (hunk 3 above)
Comment (pavelvelikhov) [reply ydb/library/yql/core/cbo/cbo_optimizer_new.h:111]: done

File: ydb/library/yql/core/cbo/cbo_optimizer_new.h (hunk 5 above)
Comment (pavelvelikhov) [reply ydb/library/yql/core/cbo/cbo_optimizer_new.h:124]: I added it here to avoid to create a new DummyProvider every time.

**Review by resetius — DISMISSED** [review]
OK

**Review by resetius — APPROVED** [review]

**Pull request merged**